cache/image_cache.go

```diff
@@ -0,0 +1,90 @@
+package cache
+
+import (
+	"encoding/json"
+	"io"
+	"log"
+
+	"github.com/buildpack/lifecycle"
+	"github.com/buildpack/lifecycle/image"
+	"github.com/pkg/errors"
+)
+
+//go:generate mockgen -package testmock -destination testmock/image_factory.go github.com/buildpack/lifecycle/cache ImageFactory
+
+type ImageFactory interface {
+	NewEmptyLocal(string) image.Image
+}
+
+type ImageCache struct {
+	logger    *log.Logger
+	factory   ImageFactory
+	origImage image.Image
+	newImage  *image.LoggingImage
+}
+
+func NewImageCache(logger *log.Logger, factory ImageFactory, origImage image.Image) *ImageCache {
+	newImage := factory.NewEmptyLocal(origImage.Name())
+	return &ImageCache{
+		logger:    logger,
+		factory:   factory,
+		origImage: origImage,
+		newImage:  image.NewLoggingImage(logger, newImage),
+	}
+}
+
+func (c *ImageCache) Name() string {
+	return c.origImage.Name()
+}
+
+func (c *ImageCache) SetMetadata(metadata lifecycle.CacheMetadata) error {
+	data, err := json.Marshal(metadata)
+	if err != nil {
+		return errors.Wrap(err, "marshall metadata")
+	}
+	return c.newImage.SetLabel(lifecycle.CacheMetadataLabel, string(data))
+}
+
+func (c *ImageCache) RetrieveMetadata() (lifecycle.CacheMetadata, bool, error) {
+	contents, err := lifecycle.GetMetadata(c.origImage, lifecycle.CacheMetadataLabel, c.logger)
+	if err != nil {
+		return lifecycle.CacheMetadata{}, false, err
+	}
+
+	metadata := lifecycle.CacheMetadata{}
+	if err := json.Unmarshal([]byte(contents), &metadata); err != nil {
+		c.logger.Printf("WARNING: image '%s' has incompatible '%s' label\n", c.origImage.Name(), lifecycle.CacheMetadataLabel)
```

Comment:
I think this warning gets printed inappropriately when the origImage doesn't exist on a normal first build (in addition to a warning about the image not existing). That problem predates this PR but maybe we can fix it while we are moving this code around

Reply:
Done.